### PR TITLE
Add volume encryption checks during the disk offering change

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR fixes the issue #9190 

Encryption type change is not allowed for the root volume during the scaleVM and upgradeVM operations. Same check should be made for the ChangeOfferingVolume API.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

1. Created two volumes with disk offering types one Encryption enabled and other non-encrypted
2. Tried changing the volume disk offerings to these two volumes to a different typed encryption offerings
3. Operations failed with appropriate errors.

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by vishesh92 — APPROVED** [review]
clgtm

**Review by rohityadavcloud — APPROVED** [review]

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -2120,7 +2120,7 @@ protected void validateDiskOfferingChecks(ServiceOfferingVO currentServiceOfferi
         DiskOfferingVO newRootDiskOffering = _diskOfferingDao.findById(newServiceOffering.getDiskOfferingId());
 
         if (currentRootDiskOffering.getEncrypt() != newRootDiskOffering.getEncrypt()) {
-            throw new InvalidParameterValueException("Cannot change volume encryption type via service offering change");
+            throw new InvalidParameterValueException("Cannot change the encryption type of a volume, please check the selected compute offering");
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2123]: use method` _volumeService.validateChangeDiskOffering(...)`

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -2235,6 +2236,12 @@ private void setNewIopsLimits(VolumeVO volume, DiskOfferingVO newDiskOffering, L
         }
     }
 
+    private void validateChangeDiskOffering(DiskOfferingVO existingDiskOffering, DiskOfferingVO newDiskOffering) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2239]: define this method in _VolumeApiService_ and use the same in _UserVMManagerImpl_.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 2 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Please review the new changes @sureshanaparti

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java [hunk 3]
```diff
@@ -2794,6 +2799,16 @@ public void checkAndRepairVolumeBasedOnConfig(DataObject dataObject, Host host)
         }
     }
 
+    @Override
+    public void validateChangeDiskOfferingEncryptionType(long existingDiskOfferingId, long newDiskOfferingId) {
+        DiskOfferingVO existingDiskOffering = diskOfferingDao.findById(existingDiskOfferingId);
+        DiskOfferingVO newDiskOffering = diskOfferingDao.findById(newDiskOfferingId);
```
Comment (sureshanaparti) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:2805]: in any case, new or old offering is removed, checked before this call?

Comment (harikrishna-patnala) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:2805]: Added findbyincludingremoved() method on existing disk offering.

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by kiranchavala — APPROVED** [review]
LGTM, tested it manually



Tried to change the disk offering  for root disk and data disk from non-encrypted  disk offering to  encrypted disk offering and vice versa.

Cloudstack gave the correct exception.



Tried it even with scale vm operation , cloudstack gave the correct messgae

Screenshot 2024-06-13 at 6 06 42 PM

**Comment by kiranchavala** [comment]
@blueorangutan test

**Comment by harikrishna-patnala** [comment]
@blueorangutan test

**Pull request merged**